osf/models/files.py

```diff
@@ -789,66 +484,141 @@ def serialize(self):
-                                  save=save)
+        return self._create_child(name, File, path=path, materialized_path=materialized_path, save=save)
 
     def append_folder(self, name, path=None, materialized_path=None, save=True):
-        return self._create_child(name, FileNode.FOLDER, path=path, materialized_path=materialized_path,
-                                  save=save)
+        return self._create_child(name, Folder, path=path, materialized_path=materialized_path, save=save)
 
     def _create_child(self, name, kind, path=None, materialized_path=None, save=True):
-        child = FileNode.resolve_class(self.provider, kind)(
+        if not self.pk:
+            logger.warn('BaseFileNode._create_child caused an implicit save because you just created a child with an unsaved parent.')
+            self.save()
+        child = self._resolve_class(kind)(
             name=name,
             node=self.node,
             path=path or '/' + name,
-            parent=self.stored_object,
-            materialized_path=materialized_path or os.path.join(
-                self.materialized_path, name) + '/' if not kind else ''
-        ).wrapped()
+            parent=self,
+            materialized_path=materialized_path or
+            os.path.join(self.materialized_path, name) + '/' if kind is Folder else ''
+        )
         if save:
             child.save()
         return child
 
     def find_child_by_name(self, name, kind=2):
-        return FileNode.resolve_class(self.provider, kind).find_one(
+        # kind == 2
+        type_cls = None
+        if kind == 1:
+            # file
+            type_cls = File
+        elif kind == 0:
+            # folder
+            type_cls = Folder
+        return self._resolve_class(type_cls).find_one(
             Q('name', 'eq', name) &
-            Q('parent', 'eq', self.id)
+            Q('parent', 'eq', self)
         )
 
+    def serialize(self):
+        return self._serialize()
+
+
+class UnableToDelete(Exception):
+    pass
+
+
+class TrashedFileNode(BaseFileNode):
+    is_deleted = True
+    _provider = None
+
+    def delete(self, user=None, parent=None, save=True, deleted_on=None):
+        if isinstance(self, TrashedFileNode):  # TODO Why is this needed
+            raise UnableToDelete('You cannot delete things that are deleted.')
+
+    def restore(self, recursive=True, parent=None, save=True, deleted_on=None):
+        """
+        Restore a file or folder
+        :param recursive:
+        :param parent:
+        :param save:
+        :param deleted_on:
+        :return:
+        """
+        type_cls = File if self.is_file else Folder
+
+        self.recast(self._resolve_class(type_cls)._typedmodels_type)
+
+        if save:
+            self.save()
+
+        if self.parent and self.parent.is_deleted:
+            raise ValueError('No parent to restore to')
+
+        return self
+
+
+class TrashedFile(TrashedFileNode):
```

Comment:
Not an issue but I feel the need to make a note.
It's odd that `TrashedFile` doesn't inherit from `File`